File: symbolic_dynamics/docs/papers211_215_sequence/qa/control_before_p212_semantics_terminalsource/capture.js
'use strict';
// Preserve the exact two current navigation originals before a milestone edit.
const fs=require('fs'),path=require('path'),crypto=require('crypto'),a=require('assert/strict'),cp=require('child_process');
const ROOT='/root/autodl-tmp/symbolic_dynamics',OWN=ROOT+'/docs/papers211_215_sequence/qa/control_before_p212_semantics_terminalsource';
const ENV={PATH:'/usr/bin:/bin',LANG:'C.UTF-8',LC_ALL:'C.UTF-8',TZ:'UTC'};
const names=['docs/papers211_215_sequence/PIPELINE_STATE.md','SYMBOLIC_DYNAMICS_STATE.md'];
const pin=b=>({bytes:b.length,sha256:crypto.createHash('sha256').update(b).digest('hex')});
const before=Object.fromEntries(names.map(n=>[n,pin(fs.readFileSync(ROOT+'/'+n))]));
a.ok(!fs.existsSync(OWN+'/originals'));fs.mkdirSync(OWN+'/originals',{mode:0o700});
const records=[],mapping=[];
for(const [i,n]of names.entries()){
  const source=ROOT+'/'+n,target=OWN+'/originals/'+path.basename(n);
  for(const [role,argv]of [['copy',['/usr/bin/cp','--no-clobber','--',source,target]],['compare',['/usr/bin/cmp','--',source,target]]]){
    const r=cp.spawnSync(argv[0],argv.slice(1),{cwd:ROOT,env:ENV,timeout:30000,maxBuffer:1024*1024});
    const receipt={argv,cwd:ROOT,environment:ENV,exit_code:r.status,signal:r.signal,error:r.error?String(r.error):null,stdout:pin(r.stdout||Buffer.alloc(0)),stderr:pin(r.stderr||Buffer.alloc(0))};
    fs.writeFileSync(OWN+'/'+i+'_'+role+'.stdout.raw',r.stdout||Buffer.alloc(0),{flag:'wx'});fs.writeFileSync(OWN+'/'+i+'_'+role+'.stderr.raw',r.stderr||Buffer.alloc(0),{flag:'wx'});records.push(receipt);
    fs.writeFileSync(OWN+'/'+i+'_'+role+'.NATIVE.json',JSON.stringify(receipt,null,2)+'\n',{flag:'wx'});a.equal(r.status,0);a.equal(r.signal,null);a.ok(!r.error);a.equal(r.stderr.length,0);a.equal(r.stdout.length,0);
  }
  a.deepEqual(pin(fs.readFileSync(source)),before[n]);a.deepEqual(pin(fs.readFileSync(target)),before[n]);a.ok(fs.readFileSync(source).equals(fs.readFileSync(target)));
  mapping.push({original:source,physical_copy:target,...before[n]});
}
fs.writeFileSync(OWN+'/MAPPING.json',JSON.stringify({status:'PHYSICAL_HISTORICAL_CONTROLS_BEFORE_P212_SEMANTICS_TERMINAL_SOURCE_ACCEPTANCE',mapping,actual_native_commands:records.length},null,2)+'\n',{flag:'wx'});
console.log(JSON.stringify({status:'PASS_EXACT_TWO_CONTROL_COPIES_AND_FOUR_NATIVE_COMMANDS',mapping,native_commands:records.length}));

File: docs/papers211_215_sequence/scouting/root_reception/ordered_algebra_memory/control_before_reception/capture.py
"""Preserve exact two live navigation controls before a documentary refresh."""
from hashlib import sha256
import json
import os
from pathlib import Path
import subprocess
import time

ROOT = Path('/root/autodl-tmp/symbolic_dynamics')
OUT = Path(__file__).resolve().parent
ENV = {'PATH': '/usr/bin:/bin', 'LANG': 'C.UTF-8', 'LC_ALL': 'C.UTF-8', 'TZ': 'UTC'}


def pin(data):
    return {'bytes': len(data), 'sha256': sha256(data).hexdigest()}


def put(path, value):
    body = value if isinstance(value, bytes) else (json.dumps(value, sort_keys=True, indent=2)+'\n').encode()
    with path.open('xb') as stream:
        stream.write(body)


assert OUT == ROOT/'docs/papers211_215_sequence/scouting/root_reception/ordered_algebra_memory/control_before_reception'
originals = OUT/'originals'
assert not os.path.lexists(originals)
originals.mkdir()
records, mapping = [], []
for source in (ROOT/'SYMBOLIC_DYNAMICS_STATE.md', ROOT/'docs/papers211_215_sequence/PIPELINE_STATE.md'):
    data = source.read_bytes()
    target = originals/source.name
    assert not os.path.lexists(target)
    for label, argv in (('copy_'+source.name, ['/usr/bin/cp', '--', str(source), str(target)]),
                        ('cmp_'+source.name, ['/usr/bin/cmp', '--', str(source), str(target)])):
        request = {'label': label, 'argv': argv, 'cwd': str(ROOT), 'environment': ENV,
                   'stdin': 'DEVNULL', 'timeout_seconds': 30, 'started_epoch': time.time()}
        put(OUT/(label+'.ATTEMPT.json'), request)
        result = subprocess.run(argv, cwd=ROOT, env=ENV, stdin=subprocess.DEVNULL,
                                stdout=subprocess.PIPE, stderr=subprocess.PIPE, timeout=30)
        put(OUT/(label+'.stdout.raw'), result.stdout)
        put(OUT/(label+'.stderr.raw'), result.stderr)
        receipt = {**request, 'ended_epoch': time.time(), 'exit_code': result.returncode,
                   'stdout': pin(result.stdout), 'stderr': pin(result.stderr)}
        put(OUT/(label+'.RECEIPT.json'), receipt)
        records.append(receipt)
        assert result.returncode == 0 and result.stdout == result.stderr == b''
    assert source.read_bytes() == target.read_bytes() == data
    mapping.append({'logical_path': str(source), 'physical_original': str(target), 'pin': pin(data)})
put(OUT/'MAPPING.json', mapping)
result = {'status': 'EXACT_TWO_PRE_41_CLOSED_CONTROLS_PHYSICALLY_PRESERVED',
          'actual_native_commands': records, 'mapping': mapping, 'scientific_executions': 0}
put(OUT/'RESULT.json', result)
print(json.dumps(result, sort_keys=True))
